### DISCUSSION TITLE
How do I prevent this false positive and very long events from reflected clouds?

### DISCUSSION THREAD

**Discussion by ediguidedog** [blakeblackshear/frigate#3569] [Ask A Question]
I get events like this and they last for ages. The one below has been running for about 45 minutes and will probably keep running for a bit longer yet. I don't want to add a person mask over the windscreen area as, when the car is not in the drive, that's exactly where people will be standing. I've tried adding a motion mask over the area occupied by the windscreen but from what I think I understand by the docs that only really prevents frigate initiating an object detection if there is movement in that area. Any movement outside the area, such as a passing person or car, will still trigger the detection and the whole frame is scanned including any motion masked areas. It then seems once detection is triggered some specific cloud arrangement will give a false positive and once frigate has decided a person is there the event continues as long as motion is detected somewhere in the frame (including the previously masked windscreen). 

Is the above correct and what are the recommended solutions?

Thanks 

![drive-1659359774 290128-hwu1cb](https://user-images.githubusercontent.com/73383390/182177629-de9d59a6-8fb7-413d-8a68-8e8ea51fe59b.jpg)

**Comment by NickM-27** [comment]
Yes, it seems you understand it well. In this case you'd maybe want to look at either a `min_area` filter since odds are a person worth tracking would be bigger. 

You could also just setup `detect -> stationary -> interval` which would run detection on stationary objects at that many frames. This is disabled by default but would continue to challenge stationary objects so it would continue to function as expected.

If that doesn't work then worst case you could use `detect -> stationary -> max_frames -> person` which would forcibly cancel any person object stationary for that number of frames. It can however be redected after that if there's motion nearby). Keep in mind a real person that is loitering would also be forcibly removed (and potentially redetected) in that case.

A zone may not work given the positions there.

**Reply by ediguidedog** [reply]
> Yes, it seems you understand it well. In this case you'd maybe want to look at either a `min_area` filter since odds are a person worth tracking would be bigger.

I considered that but the problem there is they are smaller. Here is the postie and he occupies less than half the pixel count of my cloud person 

![image](https://user-images.githubusercontent.com/73383390/182185237-c3d7f42f-2ebc-4573-a8ec-9a20fefcaf05.png)

> If that doesn't work then worst case you could use `detect -> stationary -> max_frames -> person` which would forcibly cancel any person object stationary for that number of frames. It can however be redected after that if there's motion nearby). Keep in mind a real person that is loitering would also be forcibly removed (and potentially redetected) in that case.

That sounds like the best option. I guess that way for the false positive to continue frigate would have to find another person in the clouds to trigger a new event and hopefully that wouldn't happen too quickly. Any real person loitering would be picked up right away so the only real downside would be multiple shorter events covering one long event.  Is that right?

edit: when you say will "cancel any person object stationary" will the event itself be cancelled and not logged or will it simply be truncated after the number of seconds and then saved and written to the event log?

**Reply by NickM-27** [reply]
> That sounds like the best option. I guess that way for the false positive to continue frigate would have to find another person in the clouds to trigger a new event and hopefully that wouldn't happen too quickly. Any real person loitering would be picked up right away so the only real downside would be multiple shorter events covering one long event. Is that right?

I am sorry but I must add that I did miss an option. You could also just setup `detect -> stationary -> interval` which would run detection on stationary objects at that many frames. This is disabled by default but would continue to challenge stationary objects so it would continue to function as expected.

> edit: when you say will "cancel any person object stationary" will the event itself be cancelled and not logged or will it simply be truncated after the number of seconds and then saved and written to the event log?

Cancel was probably a poor choice of wording, like you say it will simply be as if the detected object disappeared and the event would end as normal.

**Reply by ediguidedog** [reply]
> I am sorry but I must add that I did miss an option. You could also just setup `detect -> stationary -> interval` which would run detection on stationary objects at that many frames. This is disabled by default but would continue to challenge stationary objects so it would continue to function as expected.

So this sort of confuses me

    stationary:
    # Optional: Frequency for confirming stationary objects (default: shown below)
    # When set to 0, object detection will not confirm stationary objects until movement is detected.
    # If set to 10, object detection will run to confirm the object still exists on every 10th frame.
    interval: 0

Does frigate call the coral for object detection when it spots movement and if it gets a positive object back it starts to track that object. However if that object doesn't actually move (above a threshold) then it wont actually call the coral again and just works on the premise that the object is still there and probably stationary?  So a miscategorisation of a stationary object as a person will never be rechecked and it will then linger until its either categorised as stationary (which wont happen if there is movement elsewhere in the frame) or the object itself moves (which won't normally happen if it is a false positive) which will prompt a further call to the coral where the mistake is realized and the object is seen as lost.

So to the above configuration options. Does interval force the calling of the coral with the area of the previously identified object every n frames just to ensure it's still present and a real object of interest? If it is the event will continue and if not the event is considered over. So better detection of stationary / miscategorised objects at the expense of more calls to the coral tpu. Is that correct?

**Reply by NickM-27** [reply]
> So to the above configuration options. Does interval force the calling of the coral with the area of the previously identified object every n frames just to ensure it's still present and a real object of interest? If it is the event will continue and if not the event is considered over.

You seem to have a good grasp, but I will reiterate so it is hopefully as clear as possible.

When motion occurs frigate will start tracking that object. Once the object is stationary for `detect -> stationary -> threshold` frames it will be considered stationary.

By default, once the object is stationary frigate assumes that without motion the object will not have moved so it waits until motion occurs before running detection. If when that detection is run the objects position is similar than it is considered still stationary. However, if when that detection is run the objects position is different than it is marked as non-stationary.

> So better detection of stationary / miscategorised objects at the expense of more calls to the coral tpu. Is that correct?

Generally yes, however there are cases where it is worse (hence why it is defaulted to disabled). For example, my car parked in the driveway may have a different box based on the lighting. I don't really want it continually checking since that is work that won't need to be done (there will definitely be motion whenever the car eventually leaves the driveway). But, the option is important to have in cases like yours.

**Reply by ediguidedog** [reply]
> Generally yes, however there are cases where it is worse (hence why it is defaulted to disabled). For example, my car parked in the driveway may have a different box based on the lighting. I don't really want it continually checking since that is work that won't need to be done (there will definitely be motion whenever the car eventually leaves the driveway). But, the option is important to have in cases like yours.

And that will work well as long as your car is detected as a car and it's an object type in a zone that you don't trigger an event from. However if your stationary car is miscategorised as a person, who you do want to track in that zone, and you never recheck (as its not moving) then that person will be considered to still be there you will have a never ending event until your car moves and gets properly identified as a car. By rechecking every 10 frames my person in the clouds will hopefully turn back into a car when the clouds move a bit and as the person is now seen to have left my drive to be replaced with my car the event will stop.  So this seems to be a better way than limiting the max number of frames for any given object event.

**Reply by NickM-27** [reply]
In my case (different angle than yours) there are not reflections so this wouldn't happen (I was just providing a contrasting example). 

> So this seems to be a better way than limiting the max number of frames for any given object event.

Yes, `max_frames` is an absolute brute-force last resort option (as is hopefully indicated by the comments for that option in the docs)

**Comment by cleverpal-ai** [comment]
I have the same issue as a topic starter. My street lamp gets recognized as a "person", and I put a `min_area` in configuration in order to exclude this false positive. It doesn't trigger within my zone of interest anymore, however it records rather long event videos (for this camera I have an `active_object` recording). It happens when the rain or snow falls and such recording may occur during a considerable period of time. Definitely this needs to be corrected somehow (my version is 0.12) so that if `min_area` has not been exceeded, it would not either trigger an event or even record anything in frigate backend.

**Reply by NickM-27** [reply]
sounds like you are applying the min_area to the zone when you actually want to apply it to the camera

**Reply by ediguidedog** [reply]
I gave up after weeks of trying and ended up adjusting the area to exclude most of the windscreen. It seems to work just fine. 

The car is slightly further down the drive than normal but even with it there I can count on one hand the false positives due to clouds since I made the change 

![image](https://github.com/blakeblackshear/frigate/assets/73383390/bc55f731-4129-426e-8577-60a761a0e118)

**Reply by cleverpal-ai** [reply]
> sounds like you are applying the min_area to the zone when you actually want to apply it to the camera

Yes, I apply min_area to the zone, as lamps are located closer to the camera in the zone of interest. However, I need to track persons (their area may be smaller than I set for the zone though) behind the zone either. In case I apply min_area to the whole camera, my persons detections behind the zone will be omitted.

**Reply by NickM-27** [reply]
right, in that case it will be excluded from the zone but still detected. have you made that zone required for events?
